src/NHibernate/Hql/Ast/ANTLR/Tree/BinaryLogicOperatorNode.cs

```diff
@@ -55,29 +55,78 @@ public virtual void Initialize()
 			IType lhsType = ExtractDataType( lhs );
 			IType rhsType = ExtractDataType( rhs );
 
-			if ( lhsType == null ) 
+			if (!AdjustParameterTypesForSpecificOperators(lhs,rhs,lhsType, rhsType))
 			{
-				lhsType = rhsType;
-			}
-			if ( rhsType == null ) 
-			{
-				rhsType = lhsType;
-			}
 
-			var lshExpectedTypeAwareNode = lhs as IExpectedTypeAwareNode;
-			if (lshExpectedTypeAwareNode != null)
-			{
-				lshExpectedTypeAwareNode.ExpectedType = rhsType;
-			}
-			var rshExpectedTypeAwareNode = rhs as IExpectedTypeAwareNode;
-			if (rshExpectedTypeAwareNode != null)
-			{
-				rshExpectedTypeAwareNode.ExpectedType = lhsType;
+				if (lhsType == null)
+				{
+					lhsType = rhsType;
+				}
+				if (rhsType == null)
+				{
+					rhsType = lhsType;
+				}
+
+
+				var lshExpectedTypeAwareNode = lhs as IExpectedTypeAwareNode;
+				if (lshExpectedTypeAwareNode != null)
+				{
+					lshExpectedTypeAwareNode.ExpectedType = rhsType;
+				}
+				var rshExpectedTypeAwareNode = rhs as IExpectedTypeAwareNode;
+				if (rshExpectedTypeAwareNode != null)
+				{
+					rshExpectedTypeAwareNode.ExpectedType = lhsType;
+				}
 			}
 
 			MutateRowValueConstructorSyntaxesIfNecessary( lhsType, rhsType );
 		}
 
+		private bool AdjustParameterTypesForSpecificOperators(IASTNode lhs, IASTNode rhs, IType lhsType, IType rhsType)
+		{
+			var lhsParameter = lhs as ParameterNode;
+			var rhsParameter = rhs as ParameterNode;
+			//For LIKE expressions the SqlType of the parameter should not not be controlled by the other operand.
```

Comment:
Double negation, which I think is unintentional. However, intentional or not this isn't strictly true, is it? Looks to me like the parameter type IS controlled by the other operand, except that the length aspect (only) is ignored? So this could be written less confusingly.

Also, I don't quite like the style (with regards to the new if-statement in Initialize()) of skipping a large block of code if another large block of code that does almost the same thing happens to return true. Hmm... I think what I mean is that as currently written it looks "unbalanced".

Instead, would it work to have something like `SetExpectedType(expectedTypeAwareNode, otherOperand, otherOperandType)`, and have that replace just the `lshExpectedTypeAwareNode.ExpectedType = rhsType;` line (and one more of course)? (I haven't fully analyzed this option.)